Error: error listing zones: could not read response body: %!w(<nil>)

### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

```
Terraform v1.2.7
on darwin_arm64
+ provider registry.terraform.io/cloudflare/cloudflare v3.20.0
+ provider registry.terraform.io/confluentinc/confluent v1.0.0
+ provider registry.terraform.io/elastic/ec v0.4.1
+ provider registry.terraform.io/hashicorp/aws v4.25.0
+ provider registry.terraform.io/hashicorp/azuread v2.27.0
+ provider registry.terraform.io/hashicorp/azurerm v3.14.0
+ provider registry.terraform.io/hashicorp/random v3.3.2
+ provider registry.terraform.io/hashicorp/vault v3.8.1
+ provider registry.terraform.io/mongodb/mongodbatlas v1.4.3
+ provider registry.terraform.io/strongdm/sdm v2.6.4
```

### Affected resource(s)

`cloudflare/cloudflare`
This is an issue with the terraform cloudflare provider.

### Terraform configuration files

```hcl
# Any resource, e.g.
data "cloudflare_zone" "xxx" {
  name = "xxx"
}
```


### Debug output

```
I don't think it's interesting for this issue, dummy strings:
Terraform version
Go runtime version
CLI args
created provider logger
CLI command args
provider: plugin process exited
```

### Panic output

_No response_

### Expected output

No changes. Your infrastructure matches the configuration.

### Actual output

```
╷
│ Error: error listing zones: could not read response body: %!w(<nil>)
│ 
│   with module.stack.data.cloudflare_zone.***_dev,
│   on ../../modules/***-az-stack/dns.tf line 321, in data "cloudflare_zone" "***_dev":
│  321: data "cloudflare_zone" "***_dev" {
│ 
╵
╷
│ Error: error listing zones: could not read response body: %!w(<nil>)
│ 
│   with module.stack.data.cloudflare_zone.***,
│   on ../../modules/***-az-stack/dns.tf line 336, in data "cloudflare_zone" "***":
│  336: data "cloudflare_zone" "***" {
│ 
╵
```

### Steps to reproduce

1. Run lots of terraform plan in parallel

### Additional factoids

The cloudflare provider is supposed to have a backoff API request system in place, but it does not seem to work as we get weird errors when we launch mulitple `terraform plan` in parallel on CICD systems. And because it fails ungraciously, we cannot differentiate those failures from real failures and we need to manually retry those until it does not fail anymore.

Are we doing something wrong? Or is the cloudflare provider not being a good cloudflare API client, respecting `CLOUDFLARE_RPS`, `CLOUDFLARE_RETRIES`, `CLOUDFLARE_MIN_BACKOFF` and `CLOUDFLARE_MAX_BACKOFF`?

If it was a good cloudflare API client, our understanding is that it should backoff until successful, and not reach this unexpected " could not read response body: %!w(<nil>)" error.

### References

_No response_

I had to strip it as each plan is 20MB long, here's an example:
```
2022-08-14T10:48:20.7505014Z 2022-08-14T10:48:20.750Z [DEBUG] provider.terraform-provider-cloudflare_v3.21.0: Cloudflare API Request Details:
2022-08-14T10:48:20.7505994Z ---[ REQUEST ]---------------------------------------
2022-08-14T10:48:20.7506688Z GET /client/v4/zones?name=***.com&per_page=50 HTTP/1.1
2022-08-14T10:48:20.7507261Z Host: api.cloudflare.com
2022-08-14T10:48:20.7508006Z User-Agent: terraform/1.2.5 terraform-plugin-sdk/2.10.1 terraform-provider-cloudflare/dev
2022-08-14T10:48:20.7508906Z Authorization: ***
2022-08-14T10:48:20.7509409Z Content-Type: application/json
2022-08-14T10:48:20.7509915Z Accept-Encoding: gzip
2022-08-14T10:48:20.7510190Z
2022-08-14T10:48:20.7510199Z
2022-08-14T10:48:20.7510604Z -----------------------------------------------------: timestamp=2022-08-14T10:48:20.749Z
2022-08-14T10:48:20.8756336Z 2022-08-14T10:48:20.874Z [DEBUG] provider.terraform-provider-cloudflare_v3.21.0: Cloudflare API Response Details:
2022-08-14T10:48:20.8757231Z ---[ RESPONSE ]--------------------------------------
2022-08-14T10:48:20.8757769Z HTTP/2.0 429 Too Many Requests
2022-08-14T10:48:20.8758238Z Connection: close
2022-08-14T10:48:20.8758905Z Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
2022-08-14T10:48:20.8759550Z Cf-Cache-Status: DYNAMIC
2022-08-14T10:48:20.8760064Z Cf-Ray: 73a91e19e8288157-DFW
2022-08-14T10:48:20.8760582Z Content-Type: application/json
2022-08-14T10:48:20.8761068Z Date: Sun, 14 Aug 2022 10:48:20 GMT
2022-08-14T10:48:20.8761833Z Expect-Ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
2022-08-14T10:48:20.8762485Z Expires: Sun, 25 Jan 1981 05:00:00 GMT
2022-08-14T10:48:20.8762978Z Pragma: no-cache
2022-08-14T10:48:20.8763391Z Server: cloudflare
2022-08-14T10:48:20.8764240Z Set-Cookie: __cflb=0H28vgHxwvgAQtjUGUFqYFDiSDreGJnUoLf1MaRxVWD; SameSite=Lax; path=/; expires=Sun, 14-Aug-22 13:18:21 GMT; HttpOnly
2022-08-14T10:48:20.8765345Z Set-Cookie: __cfruid=2b509f97b59b692bc9f6681a68c5b0aa232e3bb0-1660474100; path=/; domain=.api.cloudflare.com; HttpOnly; Secure; SameSite=None
2022-08-14T10:48:20.8766167Z Strict-Transport-Security: max-age=31536000
2022-08-14T10:48:20.8766724Z Vary: Accept-Encoding
2022-08-14T10:48:20.8767222Z X-Content-Type-Options: nosniff
2022-08-14T10:48:20.8767742Z X-Frame-Options: SAMEORIGIN
2022-08-14T10:48:20.8768292Z
2022-08-14T10:48:20.8768454Z {
2022-08-14T10:48:20.8768835Z  "success": false,
2022-08-14T10:48:20.8769259Z  "errors": [
2022-08-14T10:48:20.8769642Z   {
2022-08-14T10:48:20.8770032Z    "code": 10100,
2022-08-14T10:48:20.8770697Z    "message": "More than 1200 requests per 300 seconds reached. Please wait and consider throttling your request speed"
2022-08-14T10:48:20.8771305Z   }
2022-08-14T10:48:20.8771644Z  ],
2022-08-14T10:48:20.8772250Z  "messages": [],
2022-08-14T10:48:20.8772667Z  "result": null
2022-08-14T10:48:20.8773055Z }
2022-08-14T10:48:20.8773676Z -----------------------------------------------------: timestamp=2022-08-14T10:48:20.874Z
2022-08-14T10:48:20.8775778Z 2022-08-14T10:48:20.875Z [ERROR] provider.terraform-provider-cloudflare_v3.21.0: Response contains error diagnostic: @module=sdk.proto @caller=github.com/hashicorp/terraform-plugin-go@v0.12.0/tfprotov5/internal/diag/diagnostics.go:56 diagnostic_severity=ERROR diagnostic_summary="error listing zones: could not read response body: %!w(<nil>)" tf_data_source_type=cloudflare_zone tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/cloudflare/cloudflare tf_req_id=5f7901ce-88f8-f8d7-9f40-fefdf842db01 diagnostic_detail= tf_rpc=ReadDataSource timestamp=2022-08-14T10:48:20.874Z
2022-08-14T10:48:20.8777783Z 2022-08-14T10:48:20.875Z [ERROR] vertex "module.stack.data.cloudflare_zone.***" error: error listing zones: could not read response body: %!w(<nil>)
2022-08-14T10:48:20.8778906Z 2022-08-14T10:48:20.875Z [ERROR] vertex "module.stack.data.cloudflare_zone.***" error: error listing zones: could not read response body: %!w(<nil>)
2022-08-14T10:48:20.8780036Z 2022-08-14T10:48:20.875Z [ERROR] vertex "module.stack.data.cloudflare_zone.*** (expand)" error: error listing zones: could not read response body: %!w(<nil>)
2022-08-14T10:48:21.0010878Z 2022-08-14T10:48:21.000Z [DEBUG] provider.terraform-provider-cloudflare_v3.21.0: Cloudflare API Request Details:
2022-08-14T10:48:21.0013043Z ---[ REQUEST ]---------------------------------------
2022-08-14T10:48:21.0013836Z GET /client/v4/zones?name=***-dev.com&per_page=50 HTTP/1.1
2022-08-14T10:48:21.0014380Z Host: api.cloudflare.com
2022-08-14T10:48:21.0015106Z User-Agent: terraform/1.2.5 terraform-plugin-sdk/2.10.1 terraform-provider-cloudflare/dev
2022-08-14T10:48:21.0015992Z Authorization: ***
2022-08-14T10:48:21.0016495Z Content-Type: application/json
2022-08-14T10:48:21.0016991Z Accept-Encoding: gzip
2022-08-14T10:48:21.0017261Z
2022-08-14T10:48:21.0017268Z
2022-08-14T10:48:21.0017675Z -----------------------------------------------------: timestamp=2022-08-14T10:48:21.000Z
2022-08-14T10:48:21.1357986Z 2022-08-14T10:48:21.135Z [DEBUG] provider.terraform-provider-cloudflare_v3.21.0: Cloudflare API Response Details:
2022-08-14T10:48:21.1358870Z ---[ RESPONSE ]--------------------------------------
2022-08-14T10:48:21.1359391Z HTTP/2.0 429 Too Many Requests
2022-08-14T10:48:21.1359870Z Connection: close
2022-08-14T10:48:21.1360545Z Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
2022-08-14T10:48:21.1361180Z Cf-Cache-Status: DYNAMIC
2022-08-14T10:48:21.1361692Z Cf-Ray: 73a91e1b8a9866d1-DFW
2022-08-14T10:48:21.1362195Z Content-Type: application/json
2022-08-14T10:48:21.1362674Z Date: Sun, 14 Aug 2022 10:48:21 GMT
2022-08-14T10:48:21.1363457Z Expect-Ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
2022-08-14T10:48:21.1364109Z Expires: Sun, 25 Jan 1981 05:00:00 GMT
2022-08-14T10:48:21.1364602Z Pragma: no-cache
2022-08-14T10:48:21.1365022Z Server: cloudflare
2022-08-14T10:48:21.1365876Z Set-Cookie: __cflb=0H28vgHxwvgAQtjUGUFqYFDiSDreGJnUoWMgNPFT6mj; SameSite=Lax; path=/; expires=Sun, 14-Aug-22 13:18:22 GMT; HttpOnly
2022-08-14T10:48:21.1367003Z Set-Cookie: __cfruid=0ddca700750d5d803afd950ecc019836bd441742-1660474101; path=/; domain=.api.cloudflare.com; HttpOnly; Secure; SameSite=None
2022-08-14T10:48:21.1367805Z Strict-Transport-Security: max-age=31536000
2022-08-14T10:48:21.1368349Z Vary: Accept-Encoding
2022-08-14T10:48:21.1369099Z X-Content-Type-Options: nosniff
2022-08-14T10:48:21.1369614Z X-Frame-Options: SAMEORIGIN
2022-08-14T10:48:21.1369911Z
2022-08-14T10:48:21.1370060Z {
2022-08-14T10:48:21.1370423Z  "success": false,
2022-08-14T10:48:21.1370846Z  "errors": [
2022-08-14T10:48:21.1371245Z   {
2022-08-14T10:48:21.1371604Z    "code": 10100,
2022-08-14T10:48:21.1372587Z    "message": "More than 1200 requests per 300 seconds reached. Please wait and consider throttling your request speed"
2022-08-14T10:48:21.1373199Z   }
2022-08-14T10:48:21.1373549Z  ],
2022-08-14T10:48:21.1373909Z  "messages": [],
2022-08-14T10:48:21.1374312Z  "result": null
2022-08-14T10:48:21.1374703Z }
2022-08-14T10:48:21.1375316Z -----------------------------------------------------: timestamp=2022-08-14T10:48:21.135Z
2022-08-14T10:48:21.1377422Z 2022-08-14T10:48:21.135Z [ERROR] provider.terraform-provider-cloudflare_v3.21.0: Response contains error diagnostic: diagnostic_severity=ERROR diagnostic_summary="error listing zones: could not read response body: %!w(<nil>)" tf_req_id=ba3ef7f2-e145-af0f-f550-d02a95751708 @caller=github.com/hashicorp/terraform-plugin-go@v0.12.0/tfprotov5/internal/diag/diagnostics.go:56 @module=sdk.proto tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/cloudflare/cloudflare tf_rpc=ReadDataSource diagnostic_detail= tf_data_source_type=cloudflare_zone timestamp=2022-08-14T10:48:21.135Z
2022-08-14T10:48:21.1379541Z 2022-08-14T10:48:21.135Z [ERROR] vertex "module.stack.data.cloudflare_zone.***_dev" error: error listing zones: could not read response body: %!w(<nil>)
2022-08-14T10:48:21.1380693Z 2022-08-14T10:48:21.135Z [ERROR] vertex "module.stack.data.cloudflare_zone.***_dev" error: error listing zones: could not read response body: %!w(<nil>)
2022-08-14T10:48:21.1381845Z 2022-08-14T10:48:21.135Z [ERROR] vertex "module.stack.data.cloudflare_zone.***_dev (expand)" error: error listing zones: could not read response body: %!w(<nil>)
2022-08-14T11:00:28.0199532Z 2022-08-14T11:00:28.019Z [INFO]  backend/local: plan operation completed
2022-08-14T11:00:28.0240933Z [31m╷[0m[0m
2022-08-14T11:00:28.0241729Z [31m│[0m [0m[1m[31mError: [0m[0m[1merror listing zones: could not read response body: %!w(<nil>)[0m
2022-08-14T11:00:28.0242354Z [31m│[0m [0m
2022-08-14T11:00:28.0243040Z [31m│[0m [0m[0m  with module.stack.data.cloudflare_zone.***_dev,
2022-08-14T11:00:28.0243915Z [31m│[0m [0m  on ../../modules/***-az-stack/dns.tf line 321, in data "cloudflare_zone" "***_dev":
2022-08-14T11:00:28.0244803Z [31m│[0m [0m 321: data "cloudflare_zone" "***_dev" [4m{[0m[0m
2022-08-14T11:00:28.0245339Z [31m│[0m [0m
2022-08-14T11:00:28.0245776Z [31m╵[0m[0m
2022-08-14T11:00:28.0278242Z [31m╷[0m[0m
2022-08-14T11:00:28.0279521Z [31m│[0m [0m[1m[31mError: [0m[0m[1merror listing zones: could not read response body: %!w(<nil>)[0m
2022-08-14T11:00:28.0280199Z [31m│[0m [0m
2022-08-14T11:00:28.0280874Z [31m│[0m [0m[0m  with module.stack.data.cloudflare_zone.***,
2022-08-14T11:00:28.0282028Z [31m│[0m [0m  on ../../modules/***-az-stack/dns.tf line 336, in data "cloudflare_zone" "***":
2022-08-14T11:00:28.0282808Z [31m│[0m [0m 336: data "cloudflare_zone" "***" [4m{[0m[0m
2022-08-14T11:00:28.0283441Z [31m│[0m [0m
2022-08-14T11:00:28.0283872Z [31m╵[0m[0m
2022-08-14T11:00:28.0292495Z 2022-08-14T11:00:28.029Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2022-08-14T11:00:28.0299080Z 2022-08-14T11:00:28.029Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.21.0/linux_amd64/terraform-provider-cloudflare_v3.21.0 pid=2182
2022-08-14T11:00:28.0300293Z 2022-08-14T11:00:28.029Z [DEBUG] provider: plugin exited
2022-08-14T11:00:29.2079540Z 
2022-08-14T11:00:29.2528602Z 
2022-08-14T11:00:29.9896852Z 
2022-08-14T11:00:29.9923225Z ##[error]Terraform exited with code 1.
2022-08-14T11:00:29.9934785Z ##[error]Process completed with exit code 1.
```

--- a/.changelog/1043.txt
+++ b/.changelog/1043.txt
@@ -1,0 +1,3 @@
+```release-note:enhancement
+cloudflare: make it clear when the rate limit retries have been exhausted
+```

--- a/cloudflare.go
+++ b/cloudflare.go
@@ -251,6 +251,10 @@ func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, u
 		// retry if the server is rate limiting us or if it failed
 		// assumes server operations are rolled back on failure
 		if respErr != nil || resp.StatusCode == http.StatusTooManyRequests || resp.StatusCode >= 500 {
+			if resp.StatusCode == http.StatusTooManyRequests {
+				respErr = errors.New("exceeded available rate limit retries")
+			}
+
 			// if we got a valid http response, try to read body so we can reuse the connection
 			// see https://golang.org/pkg/net/http/#Client.Do
 			if respErr == nil {

